perceval/utils/density_matrix.py

```diff
@@ -355,14 +356,104 @@ def normalize(self):
+            res = dict() # result fo the form {measured FockState: (remaining density matrix, probability)
+            for key_fs, item_list in projectors.items():
+                basis = item_list[0] # FockBasis of possible measurement
+                projector = item_list[1]
+                prob = item_list[2]
+                if prob != 0:
+                    collapsed_dm = projector @ self.mat @ projector.T  # wave function collapse
+                    resulting_dm = DensityMatrix(collapsed_dm, basis)
+                    resulting_dm.normalize()
+                    res[key_fs] = (prob, resulting_dm)
+            return res
+        else:
+            sample = self.sample()[0]  # if you want to sample instead of keeping all the possibilities
+            measure, remaining = self._divide_fock_state(sample, modes)
+            basis, proj = self._construct_projector(modes, measure)
+            return DensityMatrix(proj @ self.mat @ proj.T, basis)
+
+    def _construct_projector(self, modes, fock_state) -> tuple[FockBasis, dok_array]:
+        """
+        Construct the projection operator onto the subspace of some number photons on some mode
+        """
+        if len(modes) != fock_state.m:
+            raise ValueError(f"you can't have {fock_state} in  {len(modes)} number of modes")
+
+        basis = FockBasis(self.m-fock_state.m, self.n_max-fock_state.n)
+        projector = dok_array((len(basis), self.size), dtype=float)
+
+        for i, fs in enumerate(self.reverse_index):
+            meas_fs, remain_fs = self._divide_fock_state(fs, modes)
+            if meas_fs == fock_state:
+                projector[basis[remain_fs], i] = 1
+
+        return basis, projector
+
+    def _construct_all_projectors(self, modes: list[int]) -> dict:
+        """
+        construct all the projectors associated with some modes
+        :return: a dictionary with for each measured state a list [fock_basis, projector, probability]
+        """
+        modes = list(set(modes))
+        res = dict()
+        for nb_measured_photons in range(self.n_max+1):
+            # FockBasis for the remaining density matrices
+            remaining_basis = FockBasis(self.m - len(modes), self.n_max - nb_measured_photons)
+            for measured_fs in xq.FSArray(len(modes), nb_measured_photons):
+                # initialisation of the empty projectors
+                res[measured_fs] = [remaining_basis, dok_array((len(remaining_basis), self.size)), 0]
+
+        diag_coefs = self.mat.diagonal()
+
+        for i, fs in enumerate(self.reverse_index):  # construction of the projectors
+            prob = abs(diag_coefs[i])
+            measured_fs, remaining_fs = self._divide_fock_state(fs, modes)
+            remaining_basis = res[measured_fs][0]
+            new_basis_idx = remaining_basis[remaining_fs]
+            res[measured_fs][1][new_basis_idx, i] = 1
+            res[measured_fs][2] += prob
```

Comment:
Yes, I was not sure either. I tried to think for sometime but could not find a way fast but I still suggested since you would know the code more than me